testing.py

```diff
@@ -968,6 +976,161 @@ def sample_cover_representation(self, name):
 
 
+class SearchClientForTesting(ExternalSearchIndex):
+    """When creating an index, limit it to a single shard and disable
+    replicas.
+
+    This makes search results more predictable.
+    """
+
+    def setup_index(self, new_index=None):
+        return super(SearchClientForTesting, self).setup_index(
+            new_index, number_of_shards=1, number_of_replicas=0
+        )
+
+
+class ExternalSearchTest(DatabaseTest):
+    """
+    These tests require elasticsearch to be running locally. If it's not, or there's
+    an error creating the index, the tests will pass without doing anything.
+
+    Tests for elasticsearch are useful for ensuring that we haven't accidentally broken
+    a type of search by changing analyzers or queries, but search needs to be tested manually
+    to ensure that it works well overall, with a realistic index.
+    """
+
+    def setup(self):
+        super(ExternalSearchTest, self).setup(mock_search=False)
+
+        self.integration = self._external_integration(
+            ExternalIntegration.ELASTICSEARCH,
+            goal=ExternalIntegration.SEARCH_GOAL,
+            url=u'http://localhost:9200',
+            settings={
+                ExternalSearchIndex.WORKS_INDEX_PREFIX_KEY : u'test_index',
+                ExternalSearchIndex.TEST_SEARCH_TERM_KEY : u'test_search_term',
+            }
+        )
+
+        try:
+            self.search = SearchClientForTesting(self._db)
+        except Exception as e:
+            self.search = None
+            logging.error(
+                "Unable to set up elasticsearch index, search tests will be skipped.",
+                exc_info=e
+            )
+
+    def teardown(self):
+        if self.search:
+            if self.search.works_index:
+                self.search.indices.delete(self.search.works_index, ignore=[404])
+            self.search.indices.delete('the_other_index', ignore=[404])
+            self.search.indices.delete('test_index-v100', ignore=[404])
+            self.search.indices.delete('test_index-v9999', ignore=[404])
+            ExternalSearchIndex.reset()
+        super(ExternalSearchTest, self).teardown()
+
+    def default_work(self, *args, **kwargs):
+        """Convenience method to create a work with a license pool
+        in the default collection.
+        """
+        work = self._work(
+            *args, with_license_pool=True,
+            collection=self._default_collection, **kwargs
+        )
+        work.set_presentation_ready()
+        return work
+
+
+class EndToEndSearchTest(ExternalSearchTest):
+    """Subclasses of this class set up real works in a real
+    search index and run searches against it.
+    """
+
+    def setup(self):
+        super(EndToEndSearchTest, self).setup()
+
+        # Create some works.
+        if not self.search:
+            # No search index is configured -- nothing to do.
+            return
+
+        self.populate_works()
```

Comment:
This is the only change I made when moving the class between files. Putting all the code to make the works into `populate_works` means the works can be added to the search index within `setup` -- this saves code and indent levels in every subclass of this class.